[InstCombine] Fold `(x < y) ? -1 : zext(x != y)` into `u/scmp(x,y)`

Poseydon42 opened the pull request:
This patch adds the aforementioned fold to InstCombine. This pattern is produced after naive implementations of 3-way comparison in high-level languages are transformed into LLVM IR and then optimized.

[Proofs](https://alive2.llvm.org/ce/z/w4QLq_)

This would close #99746

Poseydon42 committed:
Precommit tests

Poseydon42 committed:
[InstCombine] Fold (x [us]< y) ? -1 : (x != y) into [us]cmp(x, y)

Poseydon42 committed:
Update tests

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+// into
+//   %r = call iN @llvm.ucmp/scmp(%x, %y)
+Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+  if (!isa<ConstantInt>(SI.getTrueValue()) ||
```
Can you add some vector tests?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -4111,5 +4143,8 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     }
   }
 
+  if (auto *Instruction = foldSelectToCmp(SI))
```
It would be better to move this fold before expensive folds above (e.g., line 4066).

Poseydon42 committed:
Add support for vector types

Poseydon42 committed:
Update existing tests

Poseydon42 committed:
Move the fold up

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+// into
+//   %r = call iN @llvm.ucmp/scmp(%x, %y)
+Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+  if (!isa<Constant>(SI.getTrueValue()) ||
+      !dyn_cast<Constant>(SI.getTrueValue())->isAllOnesValue())
```
```suggestion
  if (!match(SI.getTrueValue(), m_AllOnes())
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+  Value *LHS, *RHS;
+  ICmpInst::Predicate NEPred;
+  if (!match(SI.getFalseValue(),
+             m_ZExt(m_ICmp(NEPred, m_Value(LHS), m_Value(RHS)))) ||
+      NEPred != ICmpInst::ICMP_NE)
```
```suggestion
             m_ZExt(m_SpecificICmp(ICmpInst::ICMP_NE, m_Value(LHS), m_Value(RHS)))))
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+
+  ICmpInst::Predicate LTPred;
+  if (!match(SI.getCondition(),
+             m_ICmp(LTPred, m_Specific(LHS), m_Specific(RHS))) ||
```
Please add a test with commuted ops.
```suggestion
             m_c_ICmp(LTPred, m_Specific(LHS), m_Specific(RHS))) ||
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+
+  bool IsSigned = ICmpInst::isSigned(LTPred);
+  Instruction *Result = Builder.CreateIntrinsic(
+      SI.getFalseValue()->getType(),
```
```suggestion
      SI.getType(),
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+    return nullptr;
+
+  bool IsSigned = ICmpInst::isSigned(LTPred);
+  Instruction *Result = Builder.CreateIntrinsic(
```
```
; bin/opt -passes=instcombine test.ll -S
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

define i32 @_ZNK6Symbol12fast_compareEPKS_(ptr %0, ptr %1) {
  %3 = icmp ult ptr %0, %1
  %4 = icmp ne ptr %0, %1
  %5 = zext i1 %4 to i32
  %6 = select i1 %3, i32 -1, i32 %5
  ret i32 %6
}
```
```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/lib/IR/IRBuilder.cpp:962: llvm::CallInst* llvm::IRBuilderBase::CreateIntrinsic(llvm::Type*, llvm::Intrinsic::ID, llvm::ArrayRef<llvm::Value*>, llvm::Instruction*, const llvm::Twine&): Assertion `Res == Intrinsic::MatchIntrinsicTypes_Match && TableRef.empty() && "Wrong types for intrinsic!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=instcombine reduced.ll
1.      Running pass "function(instcombine<max-iterations=1;no-use-loop-info;verify-fixpoint>)" on module "reduced.ll"
2.      Running pass "instcombine<max-iterations=1;no-use-loop-info;verify-fixpoint>" on function "_ZNK6Symbol12fast_compareEPKS_"
 #0 0x00007ffff7e0f7f2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20f7f2)
 #1 0x00007ffff7e0c6bf llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.20.0git+0x20c6bf)
 #2 0x00007ffff7e0c805 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007ffff7842520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007ffff78969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007ffff78969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007ffff78969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007ffff7842476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007ffff78287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007ffff782871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007ffff7839e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007ffff0295e59 llvm::IRBuilderBase::CreateIntrinsic(llvm::Type*, unsigned int, llvm::ArrayRef<llvm::Value*>, llvm::Instruction*, llvm::Twine const&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x295e59)
#12 0x00007ffff476af95 llvm::InstCombinerImpl::foldSelectToCmp(llvm::SelectInst&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.20.0git+0x14ff95)
#13 0x00007ffff4776a12 llvm::InstCombinerImpl::visitSelectInst(llvm::SelectInst&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.20.0git+0x15ba12)
#14 0x00007ffff467b9bd llvm::InstCombinerImpl::run() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.20.0git+0x609bd)
#15 0x00007ffff467c99c combineInstructionsOverFunction(llvm::Function&, llvm::InstructionWorklist&, llvm::AAResults*, llvm::AssumptionCache&, llvm::TargetLibraryInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::OptimizationRemarkEmitter&, llvm::BlockFrequencyInfo*, llvm::BranchProbabilityInfo*, llvm::ProfileSummaryInfo*, llvm::LoopInfo*, llvm::InstCombineOptions const&) InstructionCombining.cpp:0:0
#16 0x00007ffff467da22 llvm::InstCombinePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.20.0git+0x62a22)
#17 0x00007ffff32ce835 llvm::detail::PassModel<llvm::Function, llvm::InstCombinePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.20.0git+0xce835)
#18 0x00007ffff0351b25 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x351b25)
#19 0x00007ffff68d5c25 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.20.0git+0xd5c25)
#20 0x00007ffff0350676 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x350676)
#21 0x00007ffff68d6545 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.20.0git+0xd6545)
#22 0x00007ffff034e68a llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.20.0git+0x34e68a)
#23 0x00007ffff7f94277 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x2d277)
#24 0x00007ffff7fa0296 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.20.0git+0x39296)
#25 0x00007ffff7829d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#26 0x00007ffff7829e40 call_init ./csu/../csu/libc-start.c:128:20
#27 0x00007ffff7829e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#28 0x0000555555555095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

@nikic Do we plan to support `llvm.ucmp.i32.ptr`?

nikic replied:
Not right now, but maybe in the future (https://discourse.llvm.org/t/rfc-add-3-way-comparison-intrinsics/76685#support-other-types-6).

goldsteinn commented:
FWIW you can also handle the inverse case w/ `sext`: https://alive2.llvm.org/ce/z/k-L54N

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+// into
+//   %r = call iN @llvm.ucmp/scmp(%x, %y)
+Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+  if (!isa<Constant>(SI.getTrueValue()) ||
```
Think you could be moving the true/false values to variables and swapping based on the condition. Otherwise you won't be handling the inverse case: https://godbolt.org/z/rE7PeYMae

dtcxzyw replied:
This case is incorrect. It returns `-1` when `%y == %x`.

goldsteinn replied:
Sorry its `uge` not `ugt`: https://alive2.llvm.org/ce/z/TmFypc

Poseydon42 committed:
Full rewrite to accomodate two more folds

Poseydon42 committed:
Update tests

Poseydon42 committed:
Update preexisting tests

Poseydon42 commented:
I've rewritten the function to accomodate folding of all three patterns that were suggested. One of them had to be changed slightly since it was originally in non-cannonical form. Currently the following patterns are all folded to `ucmp(x, y)` or `scmp(x, y)`, depending on the signedness of the predicate in the condition of the select:
```
  (x < y) ? -1 : zext(x != y)
  (x > y) ? 1 : sext(x != y)
  (x >= y) ? zext(x != y) : -1
```
There's one test that existed before this PR that ended up regressing after I've updated the check lines, it's located in `Transforms/PhaseOrdering/partialord-ule.ll`. One of the selects in that test is folded to `ucmp`, but this then prohibits other optimizations functioning that would've worked on "raw" select. I'm curious to hear any suggestions about how this can be fixed. I can only think of somehow checking whether not folding to ucmp would be more beneficial than doing so, but I can't imagine that being a good solution as it would require running the whole pipeline (or at least an iteration of InstCombine) on two different versions of IR.

nikic commented:
> There's one test that existed before this PR that ended up regressing after I've updated the check lines, it's located in `Transforms/PhaseOrdering/partialord-ule.ll`. One of the selects in that test is folded to `ucmp`, but this then prohibits other optimizations functioning that would've worked on "raw" select. I'm curious to hear any suggestions about how this can be fixed. I can only think of somehow checking whether not folding to ucmp would be more beneficial than doing so, but I can't imagine that being a good solution as it would require running the whole pipeline (or at least an iteration of InstCombine) on two different versions of IR.

We need to implement any missing folds before doing the canonicalization. In this case, what is missing is basically https://alive2.llvm.org/ce/z/Z3YgSo if we view this independently of ucmp/scmp. (If we consider them together, than this would be a variant of ucmp/scmp -> icmp.)

nikic commented:
> This would close #99746

I don't think this fully addresses that issue, because it doesn't cover the other two classes of ucmp/scmp patterns (nested select and add of ext, https://alive2.llvm.org/ce/z/LV-_iD). As they are somewhat different, it makes sense to handle them separately.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,62 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+  ICmpInst::Predicate NEPred;
+  // (x < y) ? -1 : zext(x != y)
+  if (ICmpInst::isLT(Pred) && match(TV, m_AllOnes()) &&
+      match(FV, m_ZExt(m_c_ICmp(NEPred, m_Specific(LHS), m_Specific(RHS)))) &&
```
```suggestion
      match(FV, m_ZExt(m_c_SpecificICmp(ICmpInst::ICMP_NE, m_Specific(LHS), m_Specific(RHS)))) &&
```
This helper was recently added.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,62 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+  if (ICmpInst::isLT(Pred) && match(TV, m_AllOnes()) &&
+      match(FV, m_ZExt(m_c_ICmp(NEPred, m_Specific(LHS), m_Specific(RHS)))) &&
+      NEPred == ICmpInst::ICMP_NE)
+    Intrinsic = Builder.CreateIntrinsic(SI.getType(), IID, {LHS, RHS});
```
Intrinsic creation is always the same, move it into common code?

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3529,6 +3529,62 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+      NEPred == ICmpInst::ICMP_NE)
+    Intrinsic = Builder.CreateIntrinsic(SI.getType(), IID, {LHS, RHS});
+
+  // (x >= y) ? zext(x != y) : -1
```
Doesn't this case get canonicalized to `(x < y) ? -1 : zext(x != y)` as well?

Poseydon42 replied:
Yep, you're right, that leaves just two cases then.

goldsteinn commented:
> > There's one test that existed before this PR that ended up regressing after I've updated the check lines, it's located in `Transforms/PhaseOrdering/partialord-ule.ll`. One of the selects in that test is folded to `ucmp`, but this then prohibits other optimizations functioning that would've worked on "raw" select. I'm curious to hear any suggestions about how this can be fixed. I can only think of somehow checking whether not folding to ucmp would be more beneficial than doing so, but I can't imagine that being a good solution as it would require running the whole pipeline (or at least an iteration of InstCombine) on two different versions of IR.
> 
> We need to implement any missing folds before doing the canonicalization. In this case, what is missing is basically https://alive2.llvm.org/ce/z/tz24gi if we view this independently of ucmp/scmp. (If we consider them together, then this would be a variant of ucmp/scmp -> icmp.)

You might start by creating a new decomposed select matcher that also matches `{s,u}cmp` i.e `m_SelectWithCmp(CmpLHS, CmpRHS, CmpPred, TrueArm, FalseArm)` and trying to replace existing uses of `m_Select` with the generalized matcher.

Poseydon42 commented:
> > > There's one test that existed before this PR that ended up regressing after I've updated the check lines, it's located in `Transforms/PhaseOrdering/partialord-ule.ll`. One of the selects in that test is folded to `ucmp`, but this then prohibits other optimizations functioning that would've worked on "raw" select. I'm curious to hear any suggestions about how this can be fixed. I can only think of somehow checking whether not folding to ucmp would be more beneficial than doing so, but I can't imagine that being a good solution as it would require running the whole pipeline (or at least an iteration of InstCombine) on two different versions of IR.
> > 
> > 
> > We need to implement any missing folds before doing the canonicalization. In this case, what is missing is basically https://alive2.llvm.org/ce/z/tz24gi if we view this independently of ucmp/scmp. (If we consider them together, then this would be a variant of ucmp/scmp -> icmp.)
> 
> You might start by creating a new decomposed select matcher that also matches `{s,u}cmp` i.e `m_SelectWithCmp(CmpLHS, CmpRHS, CmpPred, TrueArm, FalseArm)` and trying to replace existing uses of `m_Select` with the generalized matcher.

I don't think I quite understand your approach, can you elaborate a bit more please?

Poseydon42 committed:
Address review comments

Poseydon42 committed:
Update the test with regression

Poseydon42 committed:
Format the code

Poseydon42 commented:
The regression was fixed by merging #103480.

Poseydon42 commented:
@dtcxzyw can you please rerun the regression benchmark again?

dtcxzyw commented:
@Poseydon42 We should fix regression https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1194#discussion_r1717248902 before landing this patch. As for the other minor regressions (e.g., uscmp + select/switch), I think we can leave them as todo :)

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3558,6 +3558,55 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+//   (x < y) ? -1 : zext(x != y)
+//   (x > y) ? 1 : sext(x != y)
+// Into ucmp/scmp(x, y), where signedness is determined by the signedness
+// of the comparison in the original sequence
```
```suggestion
// of the comparison in the original sequence.
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3558,6 +3558,55 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
+  // constant inside select to determine if swapping operands would be
+  // beneficial to us.
+  if ((ICmpInst::isGT(Pred) && match(TV, m_AllOnes())) ||
+      (ICmpInst::isLT(Pred) && match(TV, m_One())) || ICmpInst::isLE(Pred)) {
```
The isLE case here shouldn't be relevant anymore.

Poseydon42 committed:
Address review comments 2

nikic reviewed (approved):
LGTM

dtcxzyw reviewed (approved):
LGTM. Thank you!

davemgreen reviewed (commented):
If these are starting to generate llvm.scmp/ucmp, do they need to be careful with vectorization if that is not yet supported for cmp intrinsics?

nikic commented:
> If these are starting to generate llvm.scmp/ucmp, do they need to be careful with vectorization if that is not yet supported for cmp intrinsics?

Do you currently see vectorization for three-way comparison patterns? I wouldn't really expect these to occur in vectorizable contexts.

davemgreen commented:
> Do you currently see vectorization for three-way comparison patterns? I wouldn't really expect these to occur in vectorizable contexts.

You might be right, but all kinds of weird and wonderful things can come up at times. It would be like this, right? https://godbolt.org/z/denMv3fjf

I'm not against the patch and don't have any motivating case that needs vectorization. It might be worth adding some basic vector costmodel/isTriviallyVectorizable support at some point though.

alexey-bataev commented:
> > If these are starting to generate llvm.scmp/ucmp, do they need to be careful with vectorization if that is not yet supported for cmp intrinsics?

Looks like LLVM IR supports vector versions of these functions. Need to add them to isTriviallyVectorizable function and need to be sure that all targets support lowering for these intrinsics and correctly define the cost in TTI

asmok-g commented:
Heads-up: we're seeing comparator test failures in google. Working on reporducers.

dtcxzyw commented:
> Heads-up: we're seeing comparator test failures in google. Working on reporducers.

Can you check what is the version of llvm in your system? Does it include commit https://github.com/llvm/llvm-project/commit/85b6aac7c25f9d2a976a76045ace1e7afebb5965?

asmok-g commented:
Thanks! Looks like this fixes the miscompile. Verified some targets and will verify the rest.
